iroh-sync/src/store.rs

```diff
@@ -143,233 +142,243 @@ pub trait Store: std::fmt::Debug + Clone + Send + Sync + 'static {
-    pub fn start(&self) -> u64 {
-        match self {
-            Self::All => 0,
-            Self::From(start) => *start,
-            Self::ToInclusive(_) => 0,
-            Self::Inclusive(start, _) => *start,
-            Self::To(_) => 0,
-            Self::Exclusive(start, _) => *start,
-        }
+    /// Create a [Query::all] query filtered by a single author.
+    pub fn author(author: AuthorId) -> QueryBuilder<FlatQuery> {
+        Self::all().author(author)
     }
 
-    /// The optional end of the range, exclusive.
-    pub fn end(&self) -> Option<u64> {
-        match self {
-            Self::All => None,
-            Self::From(_) => None,
-            Self::ToInclusive(end) => end.checked_add(1),
-            Self::Inclusive(_, end) => end.checked_add(1),
-            Self::To(end) => Some(*end),
-            Self::Exclusive(_, end) => Some(*end),
-        }
+    /// Create a [Query::all] query filtered by a single key.
+    pub fn key(key: impl AsRef<[u8]>) -> QueryBuilder<FlatQuery> {
+        Self::all().key_exact(key)
     }
-}
 
-impl From<std::ops::Range<u64>> for Range {
-    fn from(value: std::ops::Range<u64>) -> Self {
-        Range::Exclusive(value.start, value.end)
+    /// Create a [Query::all] query filtered by a key prefix.
+    pub fn prefix(prefix: impl AsRef<[u8]>) -> QueryBuilder<FlatQuery> {
+        Self::all().key_prefix(prefix)
     }
 }
 
-impl From<std::ops::RangeFrom<u64>> for Range {
-    fn from(value: std::ops::RangeFrom<u64>) -> Self {
-        Range::From(value.start)
-    }
+/// Sort direction
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub enum Direction {
+    /// Sort ascending
+    #[default]
+    Asc,
+    /// Sort descending
+    Desc,
 }
 
-impl From<std::ops::RangeFull> for Range {
-    fn from(_: std::ops::RangeFull) -> Self {
-        Range::All
-    }
+#[derive(Debug, Default, Serialize, Deserialize)]
+struct LimitOffset {
+    limit: Option<u64>,
+    offset: Option<u64>,
 }
 
-impl From<std::ops::RangeInclusive<u64>> for Range {
-    fn from(value: std::ops::RangeInclusive<u64>) -> Self {
-        let (start, end) = value.into_inner();
-        Range::Inclusive(start, end)
+impl LimitOffset {
+    fn offset(&self) -> u64 {
+        self.offset.unwrap_or(0)
     }
-}
 
-impl From<std::ops::RangeToInclusive<u64>> for Range {
-    fn from(value: std::ops::RangeToInclusive<u64>) -> Self {
-        Range::ToInclusive(value.end)
+    fn limit(&self) -> Option<u64> {
+        self.limit
     }
 }
 
-impl From<std::ops::RangeTo<u64>> for Range {
-    fn from(value: std::ops::RangeTo<u64>) -> Self {
-        Range::To(value.end)
-    }
+#[derive(Debug, Serialize, Deserialize)]
+enum QueryKind {
+    Flat(FlatQuery),
+    SingleLatestPerKey(SingleLatestPerKeyQuery),
+}
+
+/// Fields by which the query can be sorted
+#[derive(Debug, Default, Serialize, Deserialize)]
+pub enum SortBy {
+    /// Sort by key.
+    Key,
+    /// Sort by author.
+    #[default]
+    Author,
```

Comment:
this is author and then key, right?

Reply:
Yup, renamed for clarity in the latest commit.